PMREMGenerator: #14469 Adding different weighting scheme for weighting of samples

akhosravian opened the pull request:
This is behind a toggle due to some issues with seams, but it works in many cases. Maybe the seam issue is obvious to someone with more familiarity @bhouston 

Here's some comparison screenshots:

Curret Method:
![Random Sampling](https://imgur.com/cPbSkqX.jpg)

With this change (fewer artifacts, the first roughness step is much blurrier, yellow tint disappears):
![Importance Sampling](https://imgur.com/KW4nJni.jpg)

Fixes #14469

Commit:
PMREMGenerator: Add option for using importance sampling in convolution
generation. This is behind a toggle due to some issues with seams, but
it works in many cases. #14469

bhouston commented:
Where was the yellow tint coming from?

On Thu, Dec 6, 2018, 7:02 PM Andrew Khosravian <notifications@github.com
wrote:

> This is behind a toggle due to some issues with seams, but it works in
> many cases. Maybe the seam issue is obvious to someone with more
> familiarity @bhouston <https://github.com/bhouston>
>
> Here's some comparison screenshots:
>
> Curret Method:
> [image: Random Sampling]
> <https://camo.githubusercontent.com/541ccdb932b18a142d46b3636b618a219f7bf62e/68747470733a2f2f696d6775722e636f6d2f635062536b71582e6a7067>
>
> With this change (fewer artifacts, the first roughness step is much
> blurrier, yellow tint disappears):
> [image: Importance Sampling]
> <https://camo.githubusercontent.com/b93bba045b830c2bcc31b1bd42309f7685d19805/68747470733a2f2f696d6775722e636f6d2f4b57346e4a6e692e6a7067>
>
> Fixes #14469 <https://github.com/mrdoob/three.js/issues/14469>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/mrdoob/three.js/pull/15390
> Commit Summary
>
>    - PMREMGenerator: Add option for using importance sampling in
>    convolution
>
> File Changes
>
>    - *M* examples/js/pmrem/PMREMGenerator.js
>    <https://github.com/mrdoob/three.js/pull/15390/files#diff-0> (368)
>
> Patch Links:
>
>    - https://github.com/mrdoob/three.js/pull/15390.patch
>    - https://github.com/mrdoob/three.js/pull/15390.diff
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mrdoob/three.js/pull/15390>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAj6_W6eshUiCEQvga86Shn2oTC04U7dks5u2bAkgaJpZM4ZHiwo>
> .
>

akhosravian commented:
The sun in the original HDR image is extremely bright (it is a day time scene after all!) and yellow tinted. 

My belief is since all samples in the current implementation are weighted equally this burning day star would skew things just a little bit yellow. I can generate the map with an external tool for comparison to verify if it is a concern.

jsantell commented on examples/js/pmrem/PMREMGenerator.js:
```diff
@@ -21,11 +19,13 @@ THREE.PMREMGenerator = ( function () {
-		this.resolution = ( resolution !== undefined ) ? resolution : 256; // NODE: 256 is currently hard coded in the glsl code for performance reasons
-		this.samplesPerLevel = ( samplesPerLevel !== undefined ) ? samplesPerLevel : 32;
+		this.sourceResolution = ( params.sourceResolution !== undefined ) ? params.sourceResolution : 512;
+		this.targetResolution = ( params.targetResolution !== undefined ) ? params.targetResolution : 256; // NODE: 256 is currently hard coded in the glsl code for performance reasons
```
I think this comment can be removed, and `targetResolution` doesn't need to be a property

bhouston replied:
Why not?   I think some people may want a low resolution if they want to have multiple maps in a large environment so that one can blend between them.

jsantell replied:
To clarify, it doesn't need to be stored on the PMREMGenerator instance and could just be a local variable -- agreed it is valuable to specify the output resolution!

jsantell commented on examples/js/pmrem/PMREMGenerator.js:
```diff
@@ -21,11 +19,13 @@ THREE.PMREMGenerator = ( function () {
 		this.sourceTexture = sourceTexture;
-		this.resolution = ( resolution !== undefined ) ? resolution : 256; // NODE: 256 is currently hard coded in the glsl code for performance reasons
-		this.samplesPerLevel = ( samplesPerLevel !== undefined ) ? samplesPerLevel : 32;
+		this.sourceResolution = ( params.sourceResolution !== undefined ) ? params.sourceResolution : 512;
```
Might be helpful to check for `resolution` for folks upgrading to this version if `sourceResolution` not found, e.g. `( params.sourceResolution !== undefined ) ? params.sourceResolution : params.resolution !== undefined ? params.resolution : 512;`

bhouston replied:
good idea.

jsantell commented on examples/js/pmrem/PMREMGenerator.js:
```diff
@@ -21,11 +19,13 @@ THREE.PMREMGenerator = ( function () {
+		this.sourceResolution = ( params.sourceResolution !== undefined ) ? params.sourceResolution : 512;
+		this.targetResolution = ( params.targetResolution !== undefined ) ? params.targetResolution : 256; // NODE: 256 is currently hard coded in the glsl code for performance reasons
+		this.samplesPerLevel = ( params.samplesPerLevel !== undefined ) ? params.samplesPerLevel : 32;
+		this.useImportanceSampling = ( params.useImportanceSampling !== undefined ) ? params.useImportanceSampling : false;
```
`new PMREMGenerator({ useImportanceSampling: false })` would not use importance sampling in this case which would be unexpected

bhouston replied:
Regard this style, I believe Three.JS does not use objects passed in as parameters, rather it uses ordered arguments.  The reason for this is that in generally objects used as parameter sets on a regular basis , because they are memory allocations on the heap, introduces GC churn.  Where as order arguments I believe are just stack based instead and thus do not impact GC.

jsantell replied:
Makes sense! Maybe a breaking API change would be useful when/if moving to importance sampling method for consistency with core (1 object allocation won't add to overhead in the PMREM case) -- that being said, `useImportanceSampling: false` should use random sampling (use case, me flipping a boolean testing cubemaps back to back with importance vs random sampling)

jsantell commented on examples/js/pmrem/PMREMGenerator.js:
```diff
@@ -59,25 +59,18 @@ THREE.PMREMGenerator = ( function () {
+		shader.uniforms = Object.assign(getBaseUniforms(), extraUniforms);
+
+		var shaderMain = this.useImportanceSampling? getImportanceSamplingFragmentShader() : getRandomSamplingShader();
+		shader.fragmentShader = getBaseFragmentShader() + shaderMain;
```
`shader` is used by all instances of PMREMGenerator -- rational was due to a memory leak (https://github.com/mrdoob/three.js/pull/15330) that was mostly from WebGLState, but some build up from creating ShaderMaterial. Shader materials with the same shaders and uniforms should roll out to the same program under the hood AFAIK, so this shouldn't be the case, and I didn't dig into it anymore.

All that being said, maybe we don't need to use the same material instance across all PMREMGenerators, or at the very least store up to 2 (random and importance) materials, since right now it'd be impossible to test both the importance PMREM alongside the random sampling version

bhouston replied:
People will generally not switch between random and importance at run time except for testing purposes (which we do not have to optimize for.)  In fact, I think that one the bugs are worked out with importance, random could potentially be removed.

jsantell replied:
@bhouston agree that switching is an unlikely use case and ultimately only need one implementation! Just the current implementation would result in args not matching what shader would be used

jsantell reviewed (commented):
Great stuff! I'm thrilled to see the snowy/sparky artifacts in the first roughness level smoothed out. Looking forward to trying this out

bhouston commented:
This is an amazing contribution.  Sorry for the earlier brief message, was traveling.

bhouston commented:
> The sun in the original HDR image is extremely bright (it is a day time scene after all!) and yellow tinted.

The sampling pattern is not wide enough to cause even the first level to have a yellow shift in the original.  I am sure of that.  There must a logical reason for the source of the yellow shift and why it went away.  I would think the issue is in regards to encode/decode somewhere along the lines.

WestLangley commented:
Thanks everyone for working on this. :-)

Here are some notes I made to myself some time ago:

1. `PMREMGenerator` constructor changes the properties of the input `sourceTexture`. IMO, we should never allow that.

2. `PMREMGenerator` `matrixFromVector( vec3 n )` appears to be a nonorthogonal transform -- _presumably_ that is not correct... I noticed this because the blurring appeared to shear the image in my example.

3. `PMREMGenerator` - `PlaneBufferGeometry( 2, 2, 0 )`

4. `PMREMCubeUVPacker` - orthographic camera top and bottom are swapped

5. `PMREMCubeUVPacker` - the computed mipmaps at each blur level do not appear to be used - at least as far as I was able to tell. In other words, the test case needs to be improved so it is clear that the algorithm is working correctly.

Assuming I am correct in my observations, (2) and (5) are the most serious.

TheophileMot reviewed (commented):
Nice work! This will be a great feature. I had some trouble getting it to work, though. The code currently has importance sampling turned off by default, but when I turn it on, the result looks strange. Here's a comparison of importance off vs. on:

![screenshot from 2018-12-18 15-13-46](https://user-images.githubusercontent.com/34871447/50182624-d66ce700-02dd-11e9-9a28-00fe75321331.png)
![screenshot from 2018-12-18 15-18-58](https://user-images.githubusercontent.com/34871447/50182632-dc62c800-02dd-11e9-832f-cf2625ffd115.png)

Also, when metalness and roughness are both set to 1, the object texture looks dimpled like a golf ball. Any ideas what's causing this?

Commit:
Merge branch 'dev' into importance_sampling

akhosravian commented:
Apologies for the dead air its been a hectic bit. Looking at fixing up this change now.

jingwenhz commented:
Hello, not sure if you already have a fix in mind, but I noticed the following when trying out the code.
It appears that there are not enough samples at the higher mip levels so rough materials look very spotty. This is consistent with what @spidersharma03 noted in his previous implementation. Sampling from the previous mip level instead of the source texture appears to improve results a lot, although this may not be strictly 'correct'.

Changing
`if ( !this.useImportanceSampling && i < 5 ) shader.uniforms[ 'envMap' ].value = this.cubeLods[ i ].texture;`
to
`if ( i < 5 ) shader.uniforms[ 'envMap' ].value = this.cubeLods[ i ].texture;`

Random sampling (existing implementation)
![random_sampling](https://user-images.githubusercontent.com/40615638/50853984-beb2dc80-1338-11e9-87ba-1706425c6e8e.png)

Importance sampling
![importance_sampling_nonlod](https://user-images.githubusercontent.com/40615638/50854014-cd998f00-1338-11e9-977d-a0aaff8e73bd.png)

Importance sampling using previous mip level
![importance_sampling_lod](https://user-images.githubusercontent.com/40615638/50854025-d2f6d980-1338-11e9-8f55-b83b7e2f647d.png)

The materials above have roughness 0.6 and look more accurate with the importance sampling, but I haven't yet tested with e.g. the MetalRoughSphere scene.

Commit:
Fixes suggested by @jsantell: correctly handling switching between
importance and random sampling implementations, and killing an unneeded
instance variable

Commit:
PMREMGenerator: add option for enabling/disabling test color map for testing purposes

akhosravian commented:
Did some digging on this there seems to be issues with seams between cube map faces. I'm at a loss for what the cause of this is though. Here's a gif I grabbed of the Khronos MetalRoughSpheres scene using 2048 samples per level (the artifacts become more pronounced with higher numbers of samples).

There's a mention in the comment at the top of the file about creating an extra pixel around each face to avoid seams. Can anyone elaborate on how that works?

jingwenhz commented:
Looking at this a bit more, it seems that the suggested change is performing mipmap-filtered importance sampling described here, is this true? https://developer.nvidia.com/gpugems/GPUGems3/gpugems3_ch20.html

But since the format of HDR files is RGBE which is not monotonic, PMREMGenerator is setting both min and mag filtering to THREE.NearestFilter. In that case, reading from the higher mips would not reduce noise. Could this be a source of error that we're seeing?

bhouston commented:
Automatic generation of mipmaps is only possibly if the encoding is monotonic otherwise it will produce garbage.

jingwenhz commented:
I see, that makes sense. But if the sampling procedure is then reading from the higher mipLevels, what kind of behavior does that result in?

bhouston commented:
Thus with our encoded RGBA values we support monotonic and non-monotonic encodings via the use of the explicit mipmapping approach used in CubeUV map.

jingwenhz commented:
Thanks for the great explanation! The example you linked is really helpful for visualizing what is going on. My only question is that the importance sampling code seems to be reading from higher mip levels of the sourceTexture when constructing the cubeLods, before they get passed to the CubeUVPacker. 

Line 385:
`rgbColor.rgb += envMapTexelToLinear(textureCube(envMap, L, mipLevel)).rgb * NdotL;`

Unless there is something else going on that I don't realize? (Sorry for the many questions! I really hope to see this feature added :) )

mrdoob commented:
#18004
